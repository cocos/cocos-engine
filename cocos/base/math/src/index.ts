/*
 Copyright (c) 2013-2016 Chukong Technologies Inc.
 Copyright (c) 2017-2023 Xiamen Yaji Software Co., Ltd.

 http://www.cocos.com

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
*/

import * as bits from './bits';

import './deprecated';
import './math-native-ext';
/**
 * Export module bits.
 */

export { bits };
export { Vec2, v2 } from './vec2';
export { Vec3, v3 } from './vec3';
export { Vec4, v4 } from './vec4';
export { Quat, quat } from './quat';
export { Mat3 } from './mat3';
export { Mat4, mat4 } from './mat4';
export { AffineTransform } from './affine-transform';
export { Size, size } from './size';
export { Rect, rect } from './rect';
export { Color, color } from './color';
export * from './utils';
export * from './type-define';
export * from './math-base';

// engine internal exports
export { preTransforms } from './mat4';
