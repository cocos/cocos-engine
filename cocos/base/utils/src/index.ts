import * as memop from './memop';
import * as js from './js';

export { js };
export { memop };
export * from './x-deprecated';
// NOTE: This is deprecated, should be removed in the future
export * from './memop';
