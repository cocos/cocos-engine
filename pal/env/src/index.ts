export declare function findCanvas (): { frame: HTMLDivElement, container: HTMLDivElement, canvas: HTMLCanvasElement };
export declare function loadJsFile (path: string): Promise<void>;
