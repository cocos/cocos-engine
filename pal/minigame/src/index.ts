import { IMiniGame } from './types';

export declare const minigame: IMiniGame;
export * from './types';