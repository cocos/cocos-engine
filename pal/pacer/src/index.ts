export declare class Pacer {
    get targetFrameRate (): number;
    set targetFrameRate (val: number);
    onTick: (() => void) | null;
    start (): void;
    stop (): void;
}
