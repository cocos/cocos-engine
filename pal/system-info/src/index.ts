// TODO: some interface need to be designed in a module called 'Application' and 'Network'
import { Feature, NetworkType, Platform, Language, OS, BrowserType, PalSystemEvent } from './enum-type';

export * from './enum-type';

declare class SystemInfo {
    public get networkType (): NetworkType;
    public get isNative (): boolean;
    public get isBrowser (): boolean;
    public get isMobile (): boolean;
    public get isLittleEndian (): boolean;
    /**
     * Platform is a spacial field which is related to the build platform you choose on the Builder Panel in Cocos Creator.
     * It may point to an OS like Android or iOS.
     * It may point to a phone brand like vivo or OPPO.
     * Also it may point to an app channel like WeChat or ByteDance.
     */
    public get platform (): Platform;
    /**
     * This is the language code typed as `Language`.
     */
    public get language (): Language;
    /**
     * This is the native value return by platform
     * The valid value can be 'zh-tw','en','en-us','fr','fr-fr','es-es' and so on.
     */
    public get nativeLanguage (): string;
    public get os (): OS;
    public get osVersion (): string;
    public get osMainVersion (): number;
    public get browserType (): BrowserType;
    public get browserVersion (): string;
    public get isXR (): boolean;
    public init (): Promise<void[]> ;
    public hasFeature (feature: Feature): boolean;
    public getBatteryLevel (): number;

    public triggerGC (): void;
    public openURL (url: string): void;
    public now (): number;
    public restartJSVM (): void;
    /*
        * Trigger to exit the game
        */
    public close (): void;
    /*
        * Free the native's resources
        */
    public exit(): void;

    on (event: PalSystemEvent, cb: (...args: any) => void, target?: any): void;
    off (event: PalSystemEvent, cb?: (...args: any) => void, target?: any): void;
    // TODO: support onError
}

export declare const systemInfo: SystemInfo;
