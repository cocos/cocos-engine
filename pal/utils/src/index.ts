/*
 Copyright (c) 2022-2023 Xiamen Yaji Software Co., Ltd.

 https://www.cocos.com/

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
*/

import { warn } from '@base/debug';
import { EDITOR } from 'internal:constants';

declare const guard: unique symbol;

type Guard = typeof guard;

/**
 * Checks if a PAL implementation module is compatible with its target interface module.
 *
 * @example
 * If you write the following in somewhere:
 *
 * ```ts
 * checkPalIntegrity<typeof import('pal-interface-module')>(
 *   withImpl<typeof import('pal-implementation-module')>());
 * ```
 *
 * you will receive a compilation error
 * if your implementation module is not fulfil the interface module.
 *
 * @note This function should be easily tree-shaken.
 */
export function checkPalIntegrity<T> (impl: T & Guard): void { /* no implementation */ }

/**
 * Utility function, see example of `checkPalIntegrity()`.
 *
 */
export function withImpl<T> (): T & Guard {
    return 0 as unknown as T & Guard;
}

/**
 * This method clones methods in minigame environment, sub as `wx`, `swan` etc. to a module called minigame.
 * @param targetObject Usually it's specified as the minigame module.
 * @param originObj Original minigame environment such as `wx`, `swan` etc.
 */
export function cloneObject (targetObject: object, originObj: object): void {
    Object.keys(originObj).forEach((key) => {
        if (typeof originObj[key] === 'function') {
            targetObject[key] = originObj[key].bind(originObj);
            return;
        }
        targetObject[key] = originObj[key];
    });
}

type MiningameAudioCallbackName = 'onPlay' | 'onPause' | 'onStop' | 'onSeek';
type InnerAudioContextPolyfillConfig = {
    [CallbackName in MiningameAudioCallbackName]: boolean;
};
/**
 * This method is to create a polyfill on minigame platform when the innerAudioContext callback doesn't work.
 * @param minigameEnv Specify the minigame enviroment such as `wx`, `swan` etc.
 * @param polyfillConfig Specify the field, if it's true, the polyfill callback will be applied.
 * @param isAsynchronous Specify whether the callback is called asynchronous.
 * @returns A polyfilled createInnerAudioContext method.
 */
export function createInnerAudioContextPolyfill (minigameEnv: any, polyfillConfig: InnerAudioContextPolyfillConfig, isAsynchronous = false) {
    return (): InnerAudioContext => {
        const audioContext: InnerAudioContext = minigameEnv.createInnerAudioContext();

        // add polyfill if onPlay method doesn't work this platform
        if (polyfillConfig.onPlay) {
            const originalPlay = audioContext.play;
            let _onPlayCB: (() => void) | null = null;
            Object.defineProperty(audioContext, 'onPlay', {
                configurable: true,
                value (cb: () => void) {
                    _onPlayCB = cb;
                },
            });
            Object.defineProperty(audioContext, 'play', {
                configurable: true,
                value () {
                    originalPlay.call(audioContext);
                    if (_onPlayCB) {
                        if (isAsynchronous) {
                            setTimeout(_onPlayCB, 0);
                        } else {
                            _onPlayCB();
                        }
                    }
                },
            });
        }

        // add polyfill if onPause method doesn't work this platform
        if (polyfillConfig.onPause) {
            const originalPause = audioContext.pause;
            let _onPauseCB: (() => void) | null = null;
            Object.defineProperty(audioContext, 'onPause', {
                configurable: true,
                value (cb: () => void) {
                    _onPauseCB = cb;
                },
            });
            Object.defineProperty(audioContext, 'pause', {
                configurable: true,
                value () {
                    originalPause.call(audioContext);
                    if (_onPauseCB) {
                        if (isAsynchronous) {
                            setTimeout(_onPauseCB, 0);
                        } else {
                            _onPauseCB();
                        }
                    }
                },
            });
        }

        // add polyfill if onStop method doesn't work on this platform
        if (polyfillConfig.onStop) {
            const originalStop = audioContext.stop;
            let _onStopCB: (() => void) | null = null;
            Object.defineProperty(audioContext, 'onStop', {
                configurable: true,
                value (cb: () => void) {
                    _onStopCB = cb;
                },
            });
            Object.defineProperty(audioContext, 'stop', {
                configurable: true,
                value () {
                    originalStop.call(audioContext);
                    if (_onStopCB) {
                        if (isAsynchronous) {
                            setTimeout(_onStopCB, 0);
                        } else {
                            _onStopCB();
                        }
                    }
                },
            });
        }

        // add polyfill if onSeeked method doesn't work on this platform
        if (polyfillConfig.onSeek) {
            const originalSeek = audioContext.seek;
            let _onSeekCB: (() => void) | null = null;
            Object.defineProperty(audioContext, 'onSeeked', {
                configurable: true,
                value (cb: () => void) {
                    _onSeekCB = cb;
                },
            });
            Object.defineProperty(audioContext, 'seek', {
                configurable: true,
                value (time: number) {
                    originalSeek.call(audioContext, time);
                    if (_onSeekCB) {
                        if (isAsynchronous) {
                            setTimeout(_onSeekCB, 0);
                        } else {
                            _onSeekCB();
                        }
                    }
                },
            });
        }

        return audioContext;
    };
}

/**
 * Compare two version, version should in pattern like 3.0.0.
 * If versionA > versionB, return number larger than 0.
 * If versionA = versionB, return number euqal to 0.
 * If versionA < versionB, return number smaller than 0.
 * @param versionA
 * @param versionB
 */
export function versionCompare (versionA: string, versionB: string): number {
    const versionRegExp = /\d+\.\d+\.\d+/;
    if (!(versionRegExp.test(versionA) && versionRegExp.test(versionB))) {
        warn('wrong format of version when compare version');
        return 0;
    }
    const versionNumbersA = versionA.split('.').map((num: string) => Number.parseInt(num));
    const versionNumbersB = versionB.split('.').map((num: string) => Number.parseInt(num));
    for (let i = 0; i < 3; ++i) {
        const numberA = versionNumbersA[i];
        const numberB = versionNumbersB[i];
        if (numberA !== numberB) {
            return numberA - numberB;
        }
    }
    return 0;
}

/**
 * A custom implementation of setTimeout that uses requestAnimationFrame.
 * @param callback The function to be executed after a delay.
 * @param delay The delay time in milliseconds.
 * @param args The arguments to be passed to the callback function.
 * @returns A unique identifier for the timer.
 */
export function setTimeoutRAF (callback: (...args: any[]) => void, delay: number, ...args: unknown[]): number {
    const start = performance.now();

    const raf = requestAnimationFrame
    || window.requestAnimationFrame
    || window.webkitRequestAnimationFrame
    || window.mozRequestAnimationFrame
    || window.oRequestAnimationFrame
    || window.msRequestAnimationFrame;

    if (EDITOR || raf === undefined || globalThis.__globalXR?.isWebXR) {
        return setTimeout(callback, delay, ...args);
    }

    const handleRAF = (): void => {
        if (performance.now() - start < delay) {
            raf(handleRAF);
        } else {
            callback(...args);
        }
    };

    return raf(handleRAF);
}

/**
 * Cancels a timer that was created using the rafTimeout function.
 * @param id A numeric ID that represents the timer to be canceled.
 * @returns Nothing.
 */
export function clearTimeoutRAF (id: number): void {
    const caf = cancelAnimationFrame
        || window.cancelAnimationFrame
        || window.cancelRequestAnimationFrame
        || window.msCancelRequestAnimationFrame
        || window.mozCancelRequestAnimationFrame
        || window.oCancelRequestAnimationFrame
        || window.webkitCancelRequestAnimationFrame
        || window.msCancelAnimationFrame
        || window.mozCancelAnimationFrame
        || window.webkitCancelAnimationFrame
        || window.ocancelAnimationFrame;
    if (EDITOR || caf === undefined || globalThis.__globalXR?.isWebXR) {
        clearTimeout(id);
    } else {
        caf(id);
    }
}

/**
 * Invoke a function in next frame.
 * @param callback The function to be invoked next frame.
 * @param p1 The first parameter passed to `callback`.
 * @param p2 The seconde parameter passed to `callback`.
 */
export function callInNextTick (callback: AnyFunction, p1?: unknown, p2?: unknown): void {
    setTimeoutRAF(() => {
        callback(p1, p2);
    }, 0);
}

/**
 * This is a web specific util to checks whether a node is a descendant of a given node, that is the node itself, one of its direct
 * children (childNodes), one of the children's direct children, and so on.
 */
export function isDescendantElementOf (refNode, otherNode): boolean {
    if (typeof refNode.contains === 'function') {
        return refNode.contains(otherNode) as boolean;
    } else if (typeof refNode.compareDocumentPosition === 'function') {
        return !!(refNode.compareDocumentPosition(otherNode) & 16);
    } else {
        let node = otherNode.parentNode;
        if (node) {
            do {
                if (node === refNode) {
                    return true;
                } else {
                    node = node.parentNode;
                }
            } while (node !== null);
        }
        return false;
    }
}

/**
 * Checks whether a node is a DOM node.
 */
export function isDomNode (node): boolean {
    if (typeof window === 'object' && typeof Node === 'function') {
        // If "TypeError: Right-hand side of 'instanceof' is not callback" is thrown,
        // it should because window.Node was overwritten.
        return node instanceof Node;
    } else {
        return !!node
            && typeof node === 'object'
            && typeof node.nodeType === 'number'
            && typeof node.nodeName === 'string';
    }
}
