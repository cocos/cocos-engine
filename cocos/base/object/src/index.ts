/*
 Copyright (c) 2022-2023 Xiamen Yaji Software Co., Ltd.

 https://www.cocos.com/

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
*/

import *  as RF from './utils/requiring-frame';

export { CCObject } from './object';
export { type EditorExtendableObject, editorExtrasTag } from './editor-extras-tag';
export { CCClass, isCCClassOrFastDefined } from './class';
export { BitMask, Enum, ccenum, ValueType } from './value-types';
export { DELIMETER, createAttrsSingle, createAttrs, attr, getClassAttrs, setClassAttr, PrimitiveType, CCInteger, CCFloat, CCBoolean, CCString } from './utils/attribute';
export type { IExposedAttributes, IAcceptableAttributes } from './utils/attribute-defines';
export { getFullFormOfProperty, doValidateMethodWithProps_DEV } from './utils/preprocess-class';
export { RF };
export { PropertyStashInternalFlag, type ClassStash, type PropertyStash } from './class-stash';
