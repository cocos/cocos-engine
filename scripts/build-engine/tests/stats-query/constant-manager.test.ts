import { ConstantManager } from '../../src/constant-manager';

const cm = new ConstantManager(__dirname);

test('generateInternalConstants', () => {
    expect(cm.genInternalConstants()).toMatchSnapshot();
});

test('generateCCEnv', () => {
    expect(cm.genCCEnv()).toMatchSnapshot();
});

test('exportStaticConstants', () => {
    expect(cm.exportStaticConstants({
        mode: 'PREVIEW',
        platform: 'WECHAT',
        flags: { DEBUG: false, SERVER_MODE: true },
    })).toMatchSnapshot();
});

test('exportDynamicConstants', () => {
    expect(cm.exportDynamicConstants({
        mode: 'BUILD',
        platform: 'WECHAT',
        flags: { DEBUG: true, },
    })).toMatchSnapshot();
});

test('genBuildTimeConstants', () => {
    const result = cm.genBuildTimeConstants({
        mode: 'TEST',
        platform: 'NATIVE',
        flags: { DEBUG: false, SERVER_MODE: true },
    });
    expect(result).toMatchInlineSnapshot(`
        Object {
          "ALIPAY": false,
          "BAIDU": false,
          "BUILD": false,
          "BYTEDANCE": false,
          "COCOSPLAY": false,
          "DEBUG": false,
          "DEV": true,
          "EDITOR": false,
          "HTML5": false,
          "HUAWEI": false,
          "JSB": true,
          "LINKSURE": false,
          "MINIGAME": false,
          "NATIVE": true,
          "NET_MODE": 0,
          "NOT_PACK_PHYSX_LIBS": false,
          "OPPO": false,
          "PREVIEW": false,
          "QTT": false,
          "RUNTIME_BASED": false,
          "SERVER_MODE": true,
          "SUPPORT_JIT": true,
          "TEST": true,
          "VIVO": false,
          "WECHAT": false,
          "XIAOMI": false,
        }
    `);
});
