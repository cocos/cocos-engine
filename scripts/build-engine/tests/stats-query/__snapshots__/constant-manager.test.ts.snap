// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exportDynamicConstants 1`] = `
"function tryDefineGlobal (name, value) {
    const _global = typeof window === 'undefined' ? global : window;
    if (typeof _global[name] === 'undefined') {
        return (_global[name] = value);
    } else {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-return
        return _global[name];
    }
}
function defined (name) {
    const _global = typeof window === 'undefined' ? global : window;
    return typeof _global[name] === 'object';
}

export const TEST = tryDefineGlobal('CC_TEST', defined('tap') || defined('QUnit'));
export const EDITOR = tryDefineGlobal('CC_EDITOR', defined('Editor') && defined('process') && ('electron' in process.versions));
export const PREVIEW = tryDefineGlobal('CC_PREVIEW', !EDITOR);
export const JSB = tryDefineGlobal('CC_JSB', defined('jsb'));
export const NATIVE = JSB;
export const HTML5 = !(EDITOR && NATIVE);
export const DEV = tryDefineGlobal('CC_DEV', true);

export const WECHAT = true;
tryDefineGlobal('CC_WECHAT', true);

export const BAIDU = false;
tryDefineGlobal('CC_BAIDU', false);

export const XIAOMI = false;
tryDefineGlobal('CC_XIAOMI', false);

export const ALIPAY = false;
tryDefineGlobal('CC_ALIPAY', false);

export const BYTEDANCE = false;
tryDefineGlobal('CC_BYTEDANCE', false);

export const OPPO = false;
tryDefineGlobal('CC_OPPO', false);

export const VIVO = false;
tryDefineGlobal('CC_VIVO', false);

export const HUAWEI = false;
tryDefineGlobal('CC_HUAWEI', false);

export const COCOSPLAY = false;
tryDefineGlobal('CC_COCOSPLAY', false);

export const QTT = false;
tryDefineGlobal('CC_QTT', false);

export const LINKSURE = false;
tryDefineGlobal('CC_LINKSURE', false);

export const BUILD = true;
tryDefineGlobal('CC_BUILD', true);

export const DEBUG = true;
tryDefineGlobal('CC_DEBUG', true);

export const MINIGAME = true;
tryDefineGlobal('CC_MINIGAME', true);

export const RUNTIME_BASED = false;
tryDefineGlobal('CC_RUNTIME_BASED', false);

export const NET_MODE = 0;

"
`;

exports[`exportStaticConstants 1`] = `
"function tryDefineGlobal (name, value) {
    const _global = typeof window === 'undefined' ? global : window;
    if (typeof _global[name] === 'undefined') {
        return (_global[name] = value);
    } else {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-return
        return _global[name];
    }
}
export const HTML5 = false;

export const NATIVE = false;

export const WECHAT = true;
tryDefineGlobal('CC_WECHAT', true);

export const BAIDU = false;
tryDefineGlobal('CC_BAIDU', false);

export const XIAOMI = false;
tryDefineGlobal('CC_XIAOMI', false);

export const ALIPAY = false;
tryDefineGlobal('CC_ALIPAY', false);

export const BYTEDANCE = false;
tryDefineGlobal('CC_BYTEDANCE', false);

export const OPPO = false;
tryDefineGlobal('CC_OPPO', false);

export const VIVO = false;
tryDefineGlobal('CC_VIVO', false);

export const HUAWEI = false;
tryDefineGlobal('CC_HUAWEI', false);

export const COCOSPLAY = false;
tryDefineGlobal('CC_COCOSPLAY', false);

export const QTT = false;
tryDefineGlobal('CC_QTT', false);

export const LINKSURE = false;
tryDefineGlobal('CC_LINKSURE', false);

export const EDITOR = false;
tryDefineGlobal('CC_EDITOR', false);

export const PREVIEW = true;
tryDefineGlobal('CC_PREVIEW', true);

export const BUILD = false;
tryDefineGlobal('CC_BUILD', false);

export const TEST = false;
tryDefineGlobal('CC_TEST', false);

export const DEBUG = false;
tryDefineGlobal('CC_DEBUG', false);

export const SERVER_MODE = true;

export const DEV = true;
tryDefineGlobal('CC_DEV', true);

export const MINIGAME = true;
tryDefineGlobal('CC_MINIGAME', true);

export const RUNTIME_BASED = false;
tryDefineGlobal('CC_RUNTIME_BASED', false);

export const SUPPORT_JIT = false;
tryDefineGlobal('CC_SUPPORT_JIT', false);

export const JSB = false;
tryDefineGlobal('CC_JSB', false);

export const NOT_PACK_PHYSX_LIBS = false;

export const NET_MODE = 0;

"
`;

exports[`generateCCEnv 1`] = `
"declare module 'cc/env'{
	/**
	 * Running in Web platform
	 */
	export const HTML5: boolean;

	/**
	 * Running in native platform (mobile app, desktop app, or simulator).
	 */
	export const NATIVE: boolean;

	/**
	 * Running in the Wechat's mini game.
	 */
	export const WECHAT: boolean;

	/**
	 * Running in the baidu's mini game.
	 */
	export const BAIDU: boolean;

	/**
	 * Running in the xiaomi's quick game.
	 */
	export const XIAOMI: boolean;

	/**
	 * Running in the alipay's mini game.
	 */
	export const ALIPAY: boolean;

	/**
	 * Running in the ByteDance's mini game.
	 */
	export const BYTEDANCE: boolean;

	/**
	 * Running in the oppo's quick game.
	 */
	export const OPPO: boolean;

	/**
	 * Running in the vivo's quick game.
	 */
	export const VIVO: boolean;

	/**
	 * Running in the huawei's quick game.
	 */
	export const HUAWEI: boolean;

	/**
	 * Running in the cocosplay.
	 */
	export const COCOSPLAY: boolean;

	/**
	 * Running in the qtt's quick game.
	 */
	export const QTT: boolean;

	/**
	 * Running in the linksure's quick game.
	 */
	export const LINKSURE: boolean;

	/**
	 * Running in the editor.
	 */
	export const EDITOR: boolean;

	/**
	 * Preview in browser or simulator.
	 */
	export const PREVIEW: boolean;

	/**
	 * Running in published project.
	 */
	export const BUILD: boolean;

	/**
	 * Running in the engine's unit test.
	 */
	export const TEST: boolean;

	/**
	 * Running debug mode.
	 */
	export const DEBUG: boolean;

	/**
	 * Running in the editor or preview.
	 */
	export const DEV: boolean;

	/**
	 * Running in mini game.
	 */
	export const MINIGAME: boolean;

	/**
	 * Running in runtime based environment.
	 */
	export const RUNTIME_BASED: boolean;

	/**
	 * Support JIT.
	 */
	export const SUPPORT_JIT: boolean;

	/**
	 * Running in environment where using JSB as the JavaScript interface binding scheme.
	 */
	export const JSB: boolean;

	/**
	 * The network access mode.
	 * - 0 Client
	 * - 1 ListenServer
	 * - 2 HostServer
	 */
	export const NET_MODE: number;

}
"
`;

exports[`generateInternalConstants 1`] = `
"declare module 'internal:constants'{
	/**
	 * Running in Web platform
	 */
	export const HTML5: boolean;

	/**
	 * Running in native platform (mobile app, desktop app, or simulator).
	 */
	export const NATIVE: boolean;

	/**
	 * Running in the Wechat's mini game.
	 */
	export const WECHAT: boolean;

	/**
	 * Running in the baidu's mini game.
	 */
	export const BAIDU: boolean;

	/**
	 * Running in the xiaomi's quick game.
	 */
	export const XIAOMI: boolean;

	/**
	 * Running in the alipay's mini game.
	 */
	export const ALIPAY: boolean;

	/**
	 * Running in the ByteDance's mini game.
	 */
	export const BYTEDANCE: boolean;

	/**
	 * Running in the oppo's quick game.
	 */
	export const OPPO: boolean;

	/**
	 * Running in the vivo's quick game.
	 */
	export const VIVO: boolean;

	/**
	 * Running in the huawei's quick game.
	 */
	export const HUAWEI: boolean;

	/**
	 * Running in the cocosplay.
	 */
	export const COCOSPLAY: boolean;

	/**
	 * Running in the qtt's quick game.
	 */
	export const QTT: boolean;

	/**
	 * Running in the linksure's quick game.
	 */
	export const LINKSURE: boolean;

	/**
	 * Running in the editor.
	 */
	export const EDITOR: boolean;

	/**
	 * Preview in browser or simulator.
	 */
	export const PREVIEW: boolean;

	/**
	 * Running in published project.
	 */
	export const BUILD: boolean;

	/**
	 * Running in the engine's unit test.
	 */
	export const TEST: boolean;

	/**
	 * Running debug mode.
	 */
	export const DEBUG: boolean;

	/**
	 * Running in the server mode.
	 */
	export const SERVER_MODE: boolean;

	/**
	 * Running in the editor or preview.
	 */
	export const DEV: boolean;

	/**
	 * Running in mini game.
	 */
	export const MINIGAME: boolean;

	/**
	 * Running in runtime based environment.
	 */
	export const RUNTIME_BASED: boolean;

	/**
	 * Support JIT.
	 */
	export const SUPPORT_JIT: boolean;

	/**
	 * Running in environment where using JSB as the JavaScript interface binding scheme.
	 */
	export const JSB: boolean;

	/**
	 * This is an internal constant to determine whether pack physx libs.
	 */
	export const NOT_PACK_PHYSX_LIBS: boolean;

	/**
	 * The network access mode.
	 * - 0 Client
	 * - 1 ListenServer
	 * - 2 HostServer
	 */
	export const NET_MODE: number;

}
"
`;
